### PR TITLE
blog: wing your way to better infrastructure with plugins

### PR DESCRIPTION
**Pull request by hasanaburayyan**
A blog post that tells the tale of a crazy heroic platform engineer named Susan.

### CODE REVIEW

File: blog/2023-02-17-plugins.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+### But Why?... 
+
+I think the "why" is important to talk about for a moment. Why should developers not care about the non-functional requirements of their application when writing code? The answer in my opinion is not that developers should not care about it, its that they don't want to care about it in most cases. Developers want to focus on innovations and pushing boundaries, not be shackled by the low level details of the cloud. This has been true through the history of software development, that we build abstraction layers on-top of implementation details. Most developers don't want to understand the inner workings of file systems and how they differ between operating systems. We just want to be able to read and write files, thus we have file system abstractions. Then if we need to handle special cases based on operating systems, or CPU architectures we expect the abstraction gives us a way to do that, without rewriting our entire code base.
+
+Though I think its worth noting that the purpose of the abstraction is not to hide implementation details and make cloud application development more vague, but rather unlock new mental models that drive innovation.
```
Comment (Chriscbr) [inline blog/2023-02-17-plugins.md:28]:
```suggestion
Though I think it's worth noting that the purpose of the abstraction is not to hide implementation details and make cloud application development more vague, but rather unlock new mental models that drive innovation. 
```

File: blog/2023-02-17-plugins.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+Though I think its worth noting that the purpose of the abstraction is not to hide implementation details and make cloud application development more vague, but rather unlock new mental models that drive innovation. 
+
+> "Being abstract is something profoundly different from being vague … The purpose of abstraction is not to be vague, but to create a new semantic level in which one can be absolutely precise" - Edsger Dijkstra
+
+This is the "why" for our plugin system in Wing, its intended to be a mechanism that supports teams to be more precise in their cloud applications. If the SDK unlocks this semantic level of thought, then the plugin system protects it.
```
Comment (Chriscbr) [inline blog/2023-02-17-plugins.md:32]:
```suggestion
This is the "why" for our plugin system in Wing, it's intended to be a mechanism that supports teams to be more precise in their cloud applications. If the SDK unlocks this semantic level of thought, then the plugin system protects it. 
```

File: blog/2023-02-17-plugins.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+### But Why?... 
+
+I think the "why" is important to talk about for a moment. Why should developers not care about the non-functional requirements of their application when writing code? The answer in my opinion is not that developers should not care about it, its that they don't want to care about it in most cases. Developers want to focus on innovations and pushing boundaries, not be shackled by the low level details of the cloud. This has been true through the history of software development, that we build abstraction layers on-top of implementation details. Most developers don't want to understand the inner workings of file systems and how they differ between operating systems. We just want to be able to read and write files, thus we have file system abstractions. Then if we need to handle special cases based on operating systems, or CPU architectures we expect the abstraction gives us a way to do that, without rewriting our entire code base.
```
Comment (Chriscbr) [inline blog/2023-02-17-plugins.md:26]:
```suggestion
I think the "why" is important to talk about for a moment. Why should developers not care about the non-functional requirements of their application when writing code? The answer in my opinion is not that developers should not care about it, it's that they don't want to care about it in most cases. Developers want to focus on innovations and pushing boundaries, not be shackled by the low level details of the cloud. This has been true through the history of software development, that we build abstraction layers on-top of implementation details. Most developers don't want to understand the inner workings of file systems and how they differ between operating systems. We just want to be able to read and write files, thus we have file system abstractions. Then if we need to handle special cases based on operating systems, or CPU architectures we expect the abstraction gives us a way to do that, without rewriting our entire code base.
```

File: blog/2023-02-17-plugins.md [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+:::note Susan Is A Fictional Character
+The out come of her success is purely speculative. Your company may not have corner offices so there is a chance you will have to just settle for the parking spot.
+:::
+
+## Ask Not What Your Plugin Can Do For You...
+
+This new plugin system is very exciting, and has a lot of possibilities. However, if it is to ever reach its full potential we need your help! If you have some ideas for useful plugins, or thoughts on additional hooks, or even just questions about how to make use of the plugin system, we want to hear from you! Open a Pull Request or an Issue on our [GitHub](https://github.com/winglang) also join our [community slack](https://t.winglang.io/slack) and let us know what you think.
```
Comment (Chriscbr) [inline blog/2023-02-17-plugins.md:125]:
```suggestion
This new plugin system is very exciting, and has a lot of possibilities. However, if it is to ever reach its full potential we need your help! If you have some ideas for useful plugins, or thoughts on additional hooks, or even just questions about how to make use of the plugin system, we want to hear from you! Open a pull request or issue on our [GitHub](https://github.com/winglang/wing) also join our [community slack](https://t.winglang.io/slack) and let us know what you think.
```

File: blog/2023-02-17-plugins.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+Want to read more about Wing plugins? checkout our [plugin documentation](https://docs.winglang.io/reference/compiler-plugins) for more information on the plugin system.
```
Comment (Chriscbr) [inline blog/2023-02-17-plugins.md:127]:
```suggestion
Want to read more about Wing plugins? Check out our [plugin documentation](https://docs.winglang.io/reference/compiler-plugins) for more information on the plugin system.
```

**Review by ShaiBer — APPROVED** [review]

File: blog/2023-02-17-plugins.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+exports.preSynth = function (app) { }
+
+// Validate that all IAM roles have a permission boundary
+exports.validate = function (config) { }
+```
+
+She plans to make use of a concept from [CDKTF](https://developer.hashicorp.com/terraform/cdktf) known as [Aspects](https://developer.hashicorp.com/terraform/cdktf/concepts/aspects) to traverse the construct tree and add the permission boundary. She can safely use this since she knows her intended target will be Terraform for AWS.
+
+```js
+var iam_role = require("@cdktf/provider-aws/lib/iam-role");
```
Comment (staycoolcall911) [inline blog/2023-02-17-plugins.md:72]:
```suggestion
const iam_role = require("@cdktf/provider-aws/lib/iam-role");
```

File: blog/2023-02-17-plugins.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+var cdktf = require("cdktf");
```
Comment (staycoolcall911) [inline blog/2023-02-17-plugins.md:73]:
```suggestion
const cdktf = require("cdktf");
```

File: blog/2023-02-17-plugins.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+exports.preSynth = function (app) {
+  if (!process.env.PERMISSION_BOUNDARY_ARN) {throw new Error("env var PERMISSION_BOUNDARY_ARN not set")}
+  cdktf.Aspects.of(app).add(new PermissionBoundaryAspect(process.env.PERMISSION_BOUNDARY_ARN))
+}
+
+// Validate that all IAM roles have a permission boundary
+exports.validate = function (config) { }
+```
+
+So above we can see she created a new Aspect class that implements the `visit` method. Each node the aspect visits will be checked to determine if it is an IAM role and if so, set the permission boundary which was passed into the plugin through an environment variable `PERMISSION_BOUNDARY_ARN`
```
Comment (staycoolcall911) [inline blog/2023-02-17-plugins.md:97]:
```suggestion
So above we can see she created a new Aspect class that implements the `visit` method. Each node the aspect visits will be checked to determine if it is an IAM role and if so, set the permission boundary which was passed into the plugin through an environment variable `PERMISSION_BOUNDARY_ARN`.
```

File: blog/2023-02-17-plugins.md [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+Finally for her validate step, she will simply traverse the Terraform config for all IAM roles and check if the permission boundary is set. Even though her `preSynth` hook will have already done it job, she knows that `preSynth` is a mutable hook and that another plugin may have altered things after.
```
Comment (staycoolcall911) [inline blog/2023-02-17-plugins.md:99]:
```suggestion
Finally for her validate step, she will simply traverse the Terraform config for all IAM roles and check if the permission boundary is set. Even though her `preSynth` hook will have already done the job, she knows that `preSynth` is a mutable hook and that another plugin may have altered things after.
```

File: blog/2023-02-17-plugins.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+// Add permission boundary to all IAM roles
+exports.preSynth = function (app) {
+  if (!process.env.PERMISSION_BOUNDARY_ARN) {throw new Error("env var PERMISSION_BOUNDARY_ARN not set")}
+  cdktf.Aspects.of(app).add(new PermissionBoundaryAspect(process.env.PERMISSION_BOUNDARY_ARN))
+}
+
+// Validate that all IAM roles have a permission boundary
+exports.validate = function (config) { }
```
Comment (staycoolcall911) [inline blog/2023-02-17-plugins.md:94]: I guess `validate` should be left out of this code snippet as it appears in the next one.

File: blog/2023-02-17-plugins.md [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+      throw new Error(`Role ${iamRole} has incorrect permission boundary. Expected: ${process.env.PERMISSION_BOUNDARY_ARN} but got: ${role.permission_boundary}}`);
+    }
+  }
+}
+```
+
+Now Susan can use the plugin in her CD pipelines to ensure that all IAM roles have the correct permission boundary set, without imposing this non-functional requirement on the application developers. Susan will go on to write more plugins to help her organization meet their security and compliance requirements. She is no longer the underappreciated platform engineer we know from the beginning of our blog post, but rather a hero with her own corner office, private parking spot, and an on call pager that never goes off.
+
+:::note Susan Is A Fictional Character
+The out come of her success is purely speculative. Your company may not have corner offices so there is a chance you will have to just settle for the parking spot.
```
Comment (staycoolcall911) [inline blog/2023-02-17-plugins.md:120]:
```suggestion
The outcome of her success is purely speculative. Your company may not have corner offices so there is a chance you will have to just settle for the parking spot.
```

File: blog/2023-02-17-plugins.md [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+:::
+
+## Ask Not What Your Plugin Can Do For You...
+
+This new plugin system is very exciting, and has a lot of possibilities. However, if it is to ever reach it full potential we need your help! If you have some ideas for useful plugins, or thoughts on additional hooks, or even just questions about how to make use of the plugin system, we want to hear from you! Open a pull request or an issue on our [GitHub](https://github.com/winglang/wing) also join our [community slack](https://t.winglang.io/slack) and let us know what you think.
```
Comment (staycoolcall911) [inline blog/2023-02-17-plugins.md:125]:
```suggestion
This new plugin system is very exciting, and has a lot of possibilities. However, if it is to ever reach its full potential we need your help! If you have some ideas for useful plugins, or thoughts on additional hooks, or even just questions about how to make use of the plugin system, we want to hear from you! Open a pull request or an issue on our [GitHub](https://github.com/winglang/wing) also join our [community slack](https://t.winglang.io/slack) and let us know what you think.
```

File: blog/2023-02-17-plugins.md [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+
+Want to read more about Wing plugins? Check out our [plugin documentation](https://docs.winglang.io/reference/compiler-plugins) for more information on the plugin system.
```
Comment (staycoolcall911) [inline blog/2023-02-17-plugins.md:127]: Perhaps you can also link to the plugin examples and invite the reader to add more.

**Review by staycoolcall911 — COMMENTED** [review]
This is awesome Hasan! You did a great job taking the reader through what plugins are and how to use them. 

I'm missing more emphasis on the message that plugins are taking Wing toolchain to the next level, making it much more powerful and adapt to tackle real-world challenges. Probably just 1 or 2 sentences to call that out will suffice, since you already did a very good job detailing Wing plugins and their applications.

**Review by ShaiBer — APPROVED** [review]

**Pull request merged**